File: lab2/water.swift
import Foundation

let input = readLine()!.split(separator: " ").map { Int($0)! }
let N = input[0]
let M = input[1]

var total = 0
var full = N
var empty = 0
var steps = 0
var history: [[Int]] = []

while full > 0 {
    var row: [Int] = []
    for _ in 0..<full { row.append(1) }
    for _ in 0..<empty { row.append(0) }
    history.append(row)
    
    total += full
    empty += full
    full = 0
    steps += 1
    
    if empty >= M && empty > 0 {
        var before: [Int] = []
        for _ in 0..<full { before.append(1) }
        for _ in 0..<empty { before.append(0) }
        history.append(before)
        
        full = empty / M
        empty = empty % M
        steps += 1
    }
}

print(total, steps)
for row in history {
    for val in row {
        print(val, terminator: " ")
    }
    print()
}

File: lab2/water.go
package main

import "fmt"

func main() {
    var N, M int
    fmt.Scan(&N, &M)

    total := 0
    full := N
    empty := 0
    steps := 0
    history := [][]int{}

    for full > 0 {
        row := []int{}
        for i := 0; i < full; i++ {
            row = append(row, 1)
        }
        for i := 0; i < empty; i++ {
            row = append(row, 0)
        }
        history = append(history, row)

        total += full
        empty += full
        full = 0
        steps++

        if empty >= M && empty > 0 {
            before := []int{}
            for i := 0; i < full; i++ {
                before = append(before, 1)
            }
            for i := 0; i < empty; i++ {
                before = append(before, 0)
            }
            history = append(history, before)

            full = empty / M
            empty = empty % M
            steps++
        }
    }

    fmt.Println(total, steps)
    for i := 0; i < len(history); i++ {
        for j := 0; j < len(history[i]); j++ {
            fmt.Print(history[i][j], " ")
        }
        fmt.Println()
    }
}
